File: tests/fsharp/OpenTelemetry.fs
namespace common.tests

open System.Diagnostics
open Azure.Monitor.OpenTelemetry.AspNetCore
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open OpenTelemetry
open OpenTelemetry.Metrics
open OpenTelemetry.Trace
open FSharpPlus
open common

[<RequireQualifiedAccess>]
module Activity =
    let fromSource (name: string) (activitySource: ActivitySource) = activitySource.StartActivity(name)

    let setTag key value (activity: Activity | null) =
        Option.ofObj activity
        |> Option.map _.SetTag(key, value)
        |> Option.defaultValue Unchecked.defaultof<Activity>

[<RequireQualifiedAccess>]
module OpenTelemetry =
    let private setDestination configuration (builder: IOpenTelemetryBuilder) =
        Configuration.getValue "APPLICATION_INSIGHTS_CONNECTION_STRING" configuration
        |> iter (fun _ ->
            match builder with
            | :? OpenTelemetryBuilder as builder -> builder.UseAzureMonitor() |> ignore
            | _ -> ())

        Configuration.getValue "OTEL_EXPORTER_OTLP_ENDPOINT" configuration
        |> iter (fun _ -> builder.UseOtlpExporter() |> ignore)

        builder

    let private configureMetrics (builder: MeterProviderBuilder) =
        builder.AddHttpClientInstrumentation()
        |> _.AddAspNetCoreInstrumentation()
        |> ignore

    let private configureTracing (builder: TracerProviderBuilder) =
        builder.SetSampler(AlwaysOnSampler())
        |> _.AddHttpClientInstrumentation()
        |> _.AddAspNetCoreInstrumentation()
        |> Cosmos.configureOpenTelemetryTracing
        |> ignore

    let configureBuilder (builder: IHostApplicationBuilder) =
        builder.Services.AddOpenTelemetry()
        |> setDestination builder.Configuration
        |> _.WithMetrics(configureMetrics)
        |> _.WithTracing(configureTracing)
        |> ignore

        builder

File: tests/csharp/OpenTelemetry.cs
using Azure.Monitor.OpenTelemetry.AspNetCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using OpenTelemetry;
using OpenTelemetry.Metrics;
using OpenTelemetry.Trace;
using System.Diagnostics;

namespace common.tests;

internal static class ActivitySourceModule
{
    public static IHostApplicationBuilder ConfigureBuilder(IHostApplicationBuilder builder, string name)
    {
        builder.Services.TryAddSingleton(_ => new ActivitySource(name));

        return builder;
    }
}

internal static class ActivityModule
{
    public static Activity? FromSource(ActivitySource source, string name) =>
        source.StartActivity(name);
}

internal static class OpenTelemetryModule
{
    public static IHostApplicationBuilder ConfigureOpenTelemetry(this IHostApplicationBuilder builder)
    {
        builder.Services
               .AddOpenTelemetry()
               .SetDestination(builder.Configuration)
               .WithMetrics(ConfigureMetrics)
               .WithTracing(ConfigureTracing);

        return builder;
    }

    private static IOpenTelemetryBuilder SetDestination(this IOpenTelemetryBuilder builder, IConfiguration configuration)
    {
        configuration.GetValue("APPLICATION_INSIGHTS_CONNECTION_STRING")
                     .Iter(connectionString =>
                     {
                         switch (builder)
                         {
                             case OpenTelemetryBuilder openTelemetryBuilder:
                                 openTelemetryBuilder.UseAzureMonitor();
                                 break;
                         }
                     });

        configuration.GetValue("OTEL_EXPORTER_OTLP_ENDPOINT")
                     .Iter(endpoint => builder.UseOtlpExporter());

        return builder;
    }

    private static void ConfigureMetrics(MeterProviderBuilder builder) =>
        builder.AddHttpClientInstrumentation()
               .AddAspNetCoreInstrumentation();

    private static void ConfigureTracing(TracerProviderBuilder builder) =>
        builder.AddHttpClientInstrumentation()
               .AddAspNetCoreInstrumentation()
               .AddSource("MyCompany.MyProduct.MyLibrary")
               .ConfigureCosmosOpenTelemetryTracing();
}